src/Worker.Extensions.DurableTask/DurableTaskFunctionsMiddleware.cs

```diff
@@ -0,0 +1,205 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Functions.Worker.Middleware;
+using Microsoft.DurableTask;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.Functions.Worker.Extensions.DurableTask;
+
+internal class DurableTaskFunctionsMiddleware : IFunctionsWorkerMiddleware
+{
+    // TODO: Need to add call to action, i.e., pointer to some documentation.
+    private const string IllegalAwaitErrorMessage =
+        "An invalid asynchronous invocation was detected. This can be caused by awaiting non-durable tasks " +
+        "in an orchestrator function's implementation or by middleware that invokes asynchronous code.";
+
+    public async Task Invoke(FunctionContext functionContext, FunctionExecutionDelegate next)
+    {
+        if (!IsOrchestrationTrigger(functionContext, out BindingMetadata? triggerMetadata))
+        {
+            await next(functionContext);
+            return;
+        }
+
+        InputBindingData<object> triggerInputData = await functionContext.BindInputAsync<object>(triggerMetadata);
+        if (triggerInputData?.Value is not string encodedOrchestratorState)
```

Comment:
this C# feature is a little fancy. Is this checking that `triggerInputData.Value` does not contain a string, but if it does contain a string then that string gets assigned to `encoded OrchestratorState`, right?

Reply:
It's effectively doing three things in one expression:
1. Declares a new local variable (called `encodedOrchestratorStatus` in this case)
2. Converts the object value into a `string`, if possible
3. Returns `true` if both the string conversion succeeded and the value is non-null.

It's a very convenient feature that helps remove boilerplate. :)